### ISSUE TITLE
What is optimized within the inner and outer resampling

### ISSUE THREAD

**Issue by ir1979** [mlr-org/mlr#1907]
[[I asked this [here](https://stats.stackexchange.com/questions/292179/whats-the-meaning-of-nested-resampling)]]

I have a feature selection and regression task in which a dataset *dataset* is provided with 64x3000 numeric data. I want to use the best features for best settings possible for a number of learners (i.e., linear regressor and decision tree). I use [mlr][1] package in R. If I understand the provided [tutorial][2] correctly, I have to use resampling (terms like *inner* and *outer* resampling) [![mlr tutorial  - nested resampling][3]][3]. What is optimized during cv2 (inner resampling) and cv3 (outer resampling) in the following sample code? 

**Code:** (dataset is not provided)

    library('mlr')
    
    set.seed(1234, "L'Ecuyer")
    
    dataset = read.csv("dataset.csv")
    
    # shuffle columns
    dataset <- dataset[, c(sample(ncol(dataset) - 1), ncol(dataset))]
    
    # try to make full rank cov matrix for linear regression
    q <- qr(dataset)
    dataset <- dataset[, q$pivot[seq(q$rank)]]
    
    
    regr.task = makeRegrTask(id = "dataset.ig", data = dataset, target = "target")
    
    rdescCV2 = makeResampleDesc("CV", iters=2)
    rdescCV3 = makeResampleDesc("CV", iters=3)
    
    inner = rdescCV2
    outer = rdescCV3
    
    
    lrns = list(
      "regr.lm"
      , makeLearner("regr.rpart", minbucket = 1)
    )
    
    measures = list(mse, rsq)
    
    for (lrn1 in lrns)
    {
      set.seed(1234, "L'Ecuyer")
      
      lrnName = ifelse(typeof(lrn1) == "list", lrn1$id, lrn1)
      if (typeof(lrn1) == "list")
      {
        lrnPars = lrn1$par.vals
      } else {
        lrnPars = list()
      }
      
      lrnName2Save = lrnName
      
      lrn = makeFilterWrapper(learner = makeLearner(cl = lrnName, par.vals = lrnPars))
      
      ps = makeParamSet(
        makeDiscreteParam("fw.abs", values = seq(3, 5, 1)),
        makeDiscreteParam("fw.method", values = c('chi.squared',
                                                  'information.gain'
        )))
      
      if ("minsplit" %in% names(getParamSet(lrn)$pars))
        ps$pars$minsplit = makeIntegerParam("minsplit", lower = 2L, upper = 3L)
      
      # try to find the best feature set and setting for each learner
      res = makeTuneWrapper(
        lrn,
        resampling = inner,
        measures = measures,
        par.set = ps,
        control = makeTuneControlGrid(),
        show.info = FALSE
      )
      
      # same indices for each learner
      set.seed(1234, "L'Ecuyer")
      
      r = resample(res, regr.task, outer, models = TRUE, measures = measures)
      
      res2 = lapply(r$models, getTuneResult)
      opt.paths = lapply(res2, function(x) as.data.frame(x$opt.path))
      optimalFeatures[[lrnName2Save]] = lapply(r$models, 
                                               function(x) getFilteredFeatures(x$learner.model$next.model))
      print(res2)
      print(opt.paths)
      print(optimalFeatures[[lrnName2Save]])
    }

**Output:**

    [Resample] cross-validation iter 1: mse.test.mean=4.95e+03,rsq.test.mean=0.222
    [Resample] cross-validation iter 2: mse.test.mean=4.03e+03,rsq.test.mean=0.497
    [Resample] cross-validation iter 3: mse.test.mean= 961,rsq.test.mean=0.765
    [Resample] Aggr. Result: mse.test.mean=3.31e+03,rsq.test.mean=0.495
    [[1]]
    Tune result:
    Op. pars: fw.abs=4; fw.method=information.gain
    mse.test.mean=2.73e+03,rsq.test.mean=0.568
    
    [[2]]
    Tune result:
    Op. pars: fw.abs=5; fw.method=information.gain
    mse.test.mean=2.9e+03,rsq.test.mean=0.383
    
    [[3]]
    Tune result:
    Op. pars: fw.abs=5; fw.method=chi.squared
    mse.test.mean=6.64e+03,rsq.test.mean=-0.0448
    
    [[1]]
      fw.abs        fw.method mse.test.mean rsq.test.mean dob eol error.message exec.time
    1      3      chi.squared      4711.697     0.3248701   1  NA          <NA>      0.36
    2      4      chi.squared      2891.273     0.5480474   2  NA          <NA>      0.33
    3      5      chi.squared      2861.078     0.5526319   3  NA          <NA>      0.31
    4      3 information.gain      2726.971     0.5631411   4  NA          <NA>      0.43
    5      4 information.gain      2726.018     0.5678868   5  NA          <NA>      0.38
    6      5 information.gain      2970.028     0.5395522   6  NA          <NA>      0.39
    
    [[2]]
      fw.abs        fw.method mse.test.mean rsq.test.mean dob eol error.message exec.time
    1      3      chi.squared      5357.465    -0.2319388   1  NA          <NA>      0.34
    2      4      chi.squared      3747.050     0.2437902   2  NA          <NA>      0.35
    3      5      chi.squared      2897.023     0.3831484   3  NA          <NA>      0.31
    4      3 information.gain      5357.465    -0.2319388   4  NA          <NA>      0.41
    5      4 information.gain      3747.050     0.2437902   5  NA          <NA>      0.42
    6      5 information.gain      2897.023     0.3831484   6  NA          <NA>      0.43
    
    [[3]]
      fw.abs        fw.method mse.test.mean rsq.test.mean dob eol error.message exec.time
    1      3      chi.squared      7593.989   -0.10557789   1  NA          <NA>      0.37
    2      4      chi.squared      6786.384   -0.02621949   2  NA          <NA>      0.33
    3      5      chi.squared      6637.264   -0.04484878   3  NA          <NA>      0.32
    4      3 information.gain      7593.989   -0.10557789   4  NA          <NA>      0.40
    5      4 information.gain      6786.384   -0.02621949   5  NA          <NA>      0.39
    6      5 information.gain      6637.264   -0.04484878   6  NA          <NA>      0.41
    
    [[1]]
    [1] "RDF065u_640" "RTp_1225"    "L2u_940"     "TIC3_182"   
    
    [[2]]
    [1] "RTp_1225"    "L2u_940"     "Mor03m_813"  "TIC3_182"    "Mor03m_2294"
    
    [[3]]
    [1] "H.046_1401"  "Mor21u_2280" "RDF065u_640" "RTp_1225"    "CIC2_1660"  
    
    [Resample] cross-validation iter 1: mse.test.mean=3.13e+03,rsq.test.mean=0.509
    [Resample] cross-validation iter 2: mse.test.mean=8.04e+03,rsq.test.mean=-0.00294
    [Resample] cross-validation iter 3: mse.test.mean=3.49e+03,rsq.test.mean=0.148
    [Resample] Aggr. Result: mse.test.mean=4.89e+03,rsq.test.mean=0.218
    [[1]]
    Tune result:
    Op. pars: fw.abs=5; fw.method=chi.squared; minsplit=3
    mse.test.mean=3.15e+03,rsq.test.mean=0.443
    
    [[2]]
    Tune result:
    Op. pars: fw.abs=5; fw.method=information.gain; minsplit=3
    mse.test.mean=3.3e+03,rsq.test.mean=0.206
    
    [[3]]
    Tune result:
    Op. pars: fw.abs=5; fw.method=chi.squared; minsplit=2
    mse.test.mean=4.33e+03,rsq.test.mean=0.368
    
    [[1]]
       fw.abs        fw.method minsplit mse.test.mean rsq.test.mean dob eol error.message exec.time
    1       3      chi.squared        2      3875.576     0.3448855   1  NA          <NA>      0.35
    2       4      chi.squared        2      4054.182     0.2971222   2  NA          <NA>      0.33
    3       5      chi.squared        2      3149.302     0.4433532   3  NA          <NA>      0.34
    4       3 information.gain        2      3351.588     0.4077916   4  NA          <NA>      0.42
    5       4 information.gain        2      3904.129     0.3151364   5  NA          <NA>      0.41
    6       5 information.gain        2      3649.004     0.3833628   6  NA          <NA>      0.39
    7       3      chi.squared        3      3875.576     0.3448855   7  NA          <NA>      0.35
    8       4      chi.squared        3      4054.182     0.2971222   8  NA          <NA>      0.35
    9       5      chi.squared        3      3149.302     0.4433532   9  NA          <NA>      0.38
    10      3 information.gain        3      3351.588     0.4077916  10  NA          <NA>      0.40
    11      4 information.gain        3      3904.129     0.3151364  11  NA          <NA>      0.41
    12      5 information.gain        3      3649.004     0.3833628  12  NA          <NA>      0.42
    
    [[2]]
       fw.abs        fw.method minsplit mse.test.mean rsq.test.mean dob eol error.message exec.time
    1       3      chi.squared        2      4846.020   -0.01409290   1  NA          <NA>      0.32
    2       4      chi.squared        2      3316.516    0.20477753   2  NA          <NA>      0.32
    3       5      chi.squared        2      3304.965    0.20643353   3  NA          <NA>      0.36
    4       3 information.gain        2      4848.166   -0.01480330   4  NA          <NA>      0.43
    5       4 information.gain        2      3316.516    0.20477753   5  NA          <NA>      0.42
    6       5 information.gain        2      3304.965    0.20643353   6  NA          <NA>      0.42
    7       3      chi.squared        3      4613.949    0.05281112   7  NA          <NA>      0.38
    8       4      chi.squared        3      3316.516    0.20477753   8  NA          <NA>      0.41
    9       5      chi.squared        3      3304.965    0.20643353   9  NA          <NA>      0.33
    10      3 information.gain        3      4795.237   -0.00721534  10  NA          <NA>      0.39
    11      4 information.gain        3      3316.516    0.20477753  11  NA          <NA>      0.38
    12      5 information.gain        3      3304.965    0.20643353  12  NA          <NA>      0.36
    
    [[3]]
       fw.abs        fw.method minsplit mse.test.mean rsq.test.mean dob eol error.message exec.time
    1       3      chi.squared        2      8346.300    -0.0896325   1  NA          <NA>      0.29
    2       4      chi.squared        2     10435.255    -0.3316064   2  NA          <NA>      0.32
    3       5      chi.squared        2      4325.461     0.3684383   3  NA          <NA>      0.30
    4       3 information.gain        2      8346.300    -0.0896325   4  NA          <NA>      0.39
    5       4 information.gain        2     10435.255    -0.3316064   5  NA          <NA>      0.39
    6       5 information.gain        2      4325.461     0.3684383   6  NA          <NA>      0.41
    7       3      chi.squared        3      8346.300    -0.0896325   7  NA          <NA>      0.36
    8       4      chi.squared        3     10435.255    -0.3316064   8  NA          <NA>      0.34
    9       5      chi.squared        3      4325.461     0.3684383   9  NA          <NA>      0.34
    10      3 information.gain        3      8346.300    -0.0896325  10  NA          <NA>      0.41
    11      4 information.gain        3     10435.255    -0.3316064  11  NA          <NA>      0.44
    12      5 information.gain        3      4325.461     0.3684383  12  NA          <NA>      0.40
    
    [[1]]
    [1] "RDF065u_640" "RTp_1225"    "L2u_940"     "Mor03m_813"  "TIC3_182"   
    
    [[2]]
    [1] "RTp_1225"    "L2u_940"     "Mor03m_813"  "TIC3_182"    "Mor03m_2294"
    
    [[3]]
    [1] "H.046_1401"  "Mor21u_2280" "RDF065u_640" "RTp_1225"    "CIC2_1660" 

  [1]: https://mlr-org.github.io/mlr-tutorial/release/html
  [2]: https://mlr-org.github.io/mlr-tutorial/release/html/nested_resampling/index.html
  [3]: https://i.stack.imgur.com/vh1sZ.png

**Comment by larskotthoff** [comment]
The measures you specify, `mse` and `rsq`.

**Comment by larskotthoff** [comment]
So the post you've linked to asks a different question. What is your actual question?

**Comment by ir1979** [comment]
In the inner resampling the best features are selected to minimize mse. What is returned from the inner resampling to the outer loop?

**Comment by larskotthoff** [comment]
The measures you specify, `mse` and `rsq`.

**Comment by ir1979** [comment]
Ok. What is changed (independent variable) to minimize sse within the inner loop? Do you mean the best features are selected during the inner resampling? So, why opt.Paths (from outer resample) has an iteration over fw.abs and fw.method while they must be tuned during the inner resampling?

**Comment by larskotthoff** [comment]
Both loops optimize the same thing. The difference is that in the inner loop, the performance is the aggregate of the performances on the folds of the cross-validation, while in the outer loop the performance is the aggregate of the performances of the inner loop iterations.

**Comment by ir1979** [comment]
Let's do a step by step approach. In the outer loop, 2/3 of the dataset is selected as outer resample train set. Then it is passed to makeTuneWrapper to find the best set of features and the best feature selection method. It is done based on 2 fold cross validation (cv2). Let's call them t1 and t2. In the inner loop, for each of t1 and t2, a number of features are selected and two models m1 and m2 are created correspondingly. The m1 and m2 are measured based on t2 and t1, respectively. The inner loop checks different tuning parameters and finds the best-aggregated test models that minimize mse. The best model (? not sure) is returned as the result of the first iteration of outer resampling. This is repeated for other two folds (cv3) in my example. Correct?

**Comment by larskotthoff** [comment]
Yes (ok, technically, not the model. but the parameters are returned from the inner loop). To quote from the tutorial:

> In the outer resampling loop, we have three pairs of training/test sets. On each of these outer training sets parameter tuning is done, thereby executing the inner resampling loop. This way, we get one set of selected hyperparameters for each outer training set. Then the learner is fitted on each outer training set using the corresponding selected hyperparameters and its performance is evaluated on the outer test sets.

**Comment by ir1979** [comment]
Ok. In my code example, opt.Paths is extracted from res2 which in turn is extracted from r$models (outer loop). So, why they include tuning steps of the inner loop?

**Comment by larskotthoff** [comment]
They don't. These are the results from the tuning iterations of the outer loop.

**Comment by ir1979** [comment]
Outer loop?! If the outer loop tunes the number of features and feature selection method in addition to the inner loop optimization of the parameters?

**Comment by larskotthoff** [comment]
The inner loop only evaluates, it doesn't optimize. Quoting again from the tutorial:

> On each of these outer training sets parameter tuning is done, thereby executing the inner resampling loop.

The inner loop evaluates the model for the tuning parameters given to it by the outer loop. Reading over your description again, I think I've interpreted it in a different way than you :) To be clear, the purpose of the inner loop is to evaluate each parameter configuration proposed by the tuning process (and there's only one, in the outer loop) in a cross-validation rather than a single train/test split.

**Comment by berndbischl** [comment]
lets make this a bit clearer:

you have 3 outer loops (CV3)
and 2 inner loops (CV2)

in each outer loop you tune your hyperparams, on the ParamSet (feasible region) you have provided (in your case fw.abs, fw.method and maybe minsplit) with the tuning algorithm you have selected (grid search).

the inner resampling strategy (cv2) is used to eval each hyper par configuration. so at the end you have 3 gridsearches going on. and mlr computes for each of these 3:
- best config x
- their perfomance y
- the optmization path for the tuning containing all evals 

does this help?

**Comment by ir1979** [comment]
> lets make this a bit clearer: ...
Do you mean in the outer loop, we first select a candidate number of features and feature selector method, create a model and then pass it to the inner loop to evaluate it. Correct? If not, would you please provide me a more step by step procedure, like my previous comment above. Thanks.

**Comment by berndbischl** [comment]
what mlr is does is standard nested resampling in combination with tuning. 
please tell we wether you have read literature on that. then we can do from there.

**Comment by berndbischl** [comment]
>> Let's do a step by step approach. 
>> In the outer loop, 2/3 of the dataset is selected as outer resample train 
set.

Correct.

>> Then it is passed to makeTuneWrapper to find the best set of features and the best feature selection method. 

Correct. the TuneWrapper now starts (in your case) a grid search on the 2/3 outer training set.


>> It is done based on 2 fold cross validation (cv2). 

Yes, the grid search uses the CV2 to asses performance now for each hyperparam x vector.


>> Let's call these two parts from cv2 as t1 and t2. In the inner loop, for each of t1 and t2, a number of features are selected and two models m1 and m2 are created correspondingly. The m1 and m2 are measured based on t2 and t1, respectively. 

Yes, correct.

>> The inner loop checks different tuning parameters and finds the best-aggregated test.mean.mse. The best model (not sure what's the meaning of the best model here, since the performance is aggregated, but how to aggregate models or selected features in m1 and m2?) is returned as the result of the first iteration of outer resampling. 

Here is your mistake. the inner loop is run to ASSESS each particular hyper param config vector x. in the grid search. it is just like when you crossval a certain hyperparam setting config of a model to evaluate it. 

I mean it is also kinda visible in the pic you posted above (which is from my paper)

**Comment by berndbischl** [comment]
i would suggest to construct a simpler example then in your case, maybe just with one learner, and one param, then really enable all logging and have a detailed look at the results, to really understand the concept behind this.

**Comment by ir1979** [comment]
Ok. reading your answer again, made it more clearer for me. Thanks.

**Comment by ir1979** [comment]
Where can I find the implementation?

**Comment by berndbischl** [comment]
> Where can I find the implementation?

huh? what do you mean?

**Comment by ir1979** [comment]
I hope I can trace the implementation code of resample and makeTuneWrapper (hopefully in C or C++) to be sure about my understanding.

**Comment by berndbischl** [comment]
errm, this is clearly an R package. And you are currently posting on github. where you could easily clone the repo and have a look at the codebase. or browse it online. 

the files are also named as the functions you call in your code.

closing now.

**Comment by ir1979** [comment]
Thank you for your help. Would you please provide a link to your paper.

**Comment by berndbischl** [comment]
http://www.compstat.statistik.uni-muenchen.de/research/

4] | B. Bischl, O. Mersmann, H. Trautmann, and C. Weihs. Resampling methods for meta-model validation with recommendations for evolutionary computation. Evolutionary Computation, 20(2):249–275, 2012. [ bib \| DOI \| http ]
-- | --

this is also good:
https://link.springer.com/chapter/10.1007%2F978-0-387-47509-7_8

**Comment by ir1979** [comment]
Thank you for your help.
